File: python/15.3Sum.py
class Solution:
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        ans=[]
        if len(nums)<3:
        	return ans

        nums.sort()

        index=-1
        indexF=-1
        cnt=0
        for i in range(len(nums)):
        	if nums[i]==0:
        		cnt+=1
        		if indexF==-1:
        			indexF=i
        	if nums[i]>0:
        		index=i
        		break
        if cnt>=3:
        	ans.append([0,0,0])
        if index==-1:
        	return ans

        if indexF==-1:
        	indexF=index-1


        dic={}
       	for num in nums:
       		dic[num]=dic.get(num,0)+1


       	for i in range(len(nums[:indexF+1])):
       		if (nums[i]==nums[i-1]):
       			continue
       		for j in range(i+1,len(nums[:indexF+1])):
       			if (j!=i+1 and nums[j]==nums[j-1]):
       				continue
       			if(dic.get(-(nums[i]+nums[j]),0)!=0):
       				ans.append([nums[i],nums[j],-(nums[i]+nums[j])])

        for i in range(index,len(nums)):
        	if (nums[i]==nums[i-1]):
        		continue
        	for j in range(i+1,len(nums)):
        		if(j!=i+1 and nums[j]==nums[j-1]):
        			continue
        		if(dic.get(-(nums[i]+nums[j]),0)!=0):
        			ans.append([nums[i],nums[j],-(nums[i]+nums[j])])

       	return ans


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
      ret = []
      if len(nums) < 3:
        return ret

      nums = sorted(nums)

      dic = dict()
      for i in range(len(nums)):
        dic[nums[i]] = i
      i = 0
      while( i < len(nums) - 1):
        j = i + 1
        while ( j < len(nums)):
          t = - (nums[i] + nums[j])
          if dic.get(t) != None and dic[t] > j:
            ret.append([nums[i], nums[j], t])
            j = dic[nums[j]]
          i = dic[nums[i]] 
          j += 1
        i += 1

      return ret
       

print(Solution().threeSum([-4,-2,-2,-2,0,1,2,2,2,3,3,4,4,6,6]))

File: cpp/15.3Sum.cpp
#include <vector>
#include <unordered_map>
#include <unordered_set>


using namespace std;


//class Solution {
//public:
//    vector<vector<int>> threeSum(vector<int>& nums) {
//        vector<vector<int>>ret;
//        if (nums.size() < 3)
//            return ret;
//
//        unordered_map<int, vector<int>> dict;
//        //unordered_map<int, vector<int>>::iterator it;
//        unordered_set<unordered_set<int>>dul_set;
//
//        for (int i=0; i<nums.size()-1; i++){
//            for (int j = i+1; j<nums.size(); j++){
//                if (dict.find(-nums[j])!= dict.end()){
//                    vector<int> indexes = dict[-nums[j]];
//                    unordered_set<int> tmpSet = {nums[indexes[0]], nums[indexes[1]], nums[j]};
//                    if (dul_set.count(tmpSet) == 0 ){
//                        dul_set.insert(tmpSet);
//                        vector<int> tmpList = {nums[indexes[0]], nums[indexes[1]], nums[j]};
//                        ret.push_back(tmpList);
//                    }
//                }
//                int key = nums[i] + nums[j];
//                vector<int> val = {i, j};
//                dict[key] = val;
//            }
//        }
//        return ret;
//    }
//};
// 上述解法超时

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector <vector<int>>  ret;
        if (nums.size() < 3)
            return ret;

        sort(nums.begin(), nums.end());
        if (nums[0] > 0 || nums[nums.size() - 1] < 0)
            return ret;

        for (int i = 0; i < nums.size()-1; i++){
            if (nums[i] > 0)
                break;
            while(i > 0 && i < nums.size()-1 && nums[i] == nums[i-1])
                i ++;
            int j = i + 1;
            int k = nums.size() - 1;
            int target = -nums[i];

            while(j < k){
                if (nums[j] + nums[k] == target){
                    vector<int>tmp = {nums[i], nums[j], nums[k]};
                    ret.push_back(tmp);
                    j++;
                    while(nums[j] == nums[j-1] && j < k) j++;
                    k--;
                    while(nums[k] == nums[k+1] && j < k) k--;
                }
                else if (nums[j] + nums[k] < target){
                    j++;
                    while(nums[j] == nums[j-1] && j < k) j++;
                }
                else{
                    k--;
                    while(nums[k] == nums[k+1] && j < k) k--;
                }

            }
        }
        return ret;
    }
};